odl/util/numerics.py

```diff
@@ -0,0 +1,152 @@
+        the points that are potentially processed multiple times.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> arr = np.ones((3, 3))
+    >>> apply_on_boundary(arr, lambda x: x / 2)
+    >>> arr
+    array([[ 0.5,  0.5,  0.5],
+           [ 0.5,  1. ,  0.5],
+           [ 0.5,  0.5,  0.5]])
+    >>>
+    >>> arr = np.ones((3, 3))
+    >>> apply_on_boundary(arr, lambda x: x / 2, only_once=False)
+    >>> arr
+    array([[ 0.25,  0.5 ,  0.25],
+           [ 0.5 ,  1.  ,  0.5 ],
+           [ 0.25,  0.5 ,  0.25]])
+    >>> arr = np.ones((3, 3))
+    >>> apply_on_boundary(arr, lambda x: x / 2, only_once=True,
+    ...                   which_boundaries=((True, False), True))
+    >>> arr
+    array([[ 0.5,  0.5,  0.5],
+           [ 0.5,  1. ,  0.5],
+           [ 0.5,  1. ,  0.5]])
+    """
+    if callable(func):
+        func = [func] * array.ndim
+    elif len(func) != array.ndim:
+        raise ValueError('sequence of functions has length {}, expected {}.'
+                         ''.format(len(func), array.ndim))
+
+    if which_boundaries is None:
+        which_boundaries = ([(True, True)] * array.ndim)
+    elif len(which_boundaries) != array.ndim:
+        raise ValueError('which_boundaries has length {}, expected {}.'
+                         ''.format(len(which_boundaries), array.ndim))
+
+    if axis_order is None:
+        axis_order = list(range(array.ndim))
+    elif len(axis_order) != array.ndim:
+        raise ValueError('axis_order has length {}, expected {}.'
+                         ''.format(len(axis_order), array.ndim))
+
+    slices = [slice(None)] * array.ndim
+    for ax, function, which in zip(axis_order, func, which_boundaries):
+        # Make slices selecting left and right
+        if only_once:
+            # Exclude first and last in the dimensions which have already
+            # been processed
+            slc_l = slices.copy()
+            slc_r = slices.copy()
+        else:
+            slc_l = [slice(None)] * array.ndim
+            slc_r = [slice(None)] * array.ndim
+
+        slc_l[ax] = 0
+        slc_r[ax] = -1
+
+        try:
+            # Tuple of functions in this axis
+            func_l, func_r = function
+        except TypeError:
+            # Single function
+            func_l = func_r = function
+
+        try:
+            # Tuple of bool
+            mod_left, mod_right = which
+        except TypeError:
+            # Single bool
+            mod_left = mod_right = which
+
+        if mod_left:
+            array[slc_l] = func_l(array[slc_l])
+            start = 1
+        else:
+            start = None
+
+        if mod_right:
+            array[slc_r] = func_r(array[slc_r])
+            end = -1
+        else:
+            end = None
+
+        slices[ax] = slice(start, end)
+
+if __name__ == '__main__':
+    from doctest import testmod, NORMALIZE_WHITESPACE
+    testmod(optionflags=NORMALIZE_WHITESPACE)
```

Comment:
Sure, put it where? I think there is already an option in the pytest config, but for manual execution you would still need this?